### PR TITLE
schema: update version regex

### PR DESCRIPTION
**Pull request by kyrofa**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] If this is a bugfix. Have you checked that there is a bug report open for the issue you are trying to fix on [bug reports](https://bugs.launchpad.net/snapcraft)?
- [x] If this is a new feature. Have you discussed the design on the [forum](https://forum.snapcraft.io)?
- [x] Have you successfully run `./runtests.sh static`?
- [x] Have you successfully run `./runtests.sh unit`?

-----

In response to https://forum.snapcraft.io/t/3974, update Snapcraft to use the same pattern for version validation as snapd and the review tools.

### CODE REVIEW

File: snapcraft/tests/unit/project_loader/test_config.py [hunk 1]
```diff
@@ -909,6 +858,67 @@ def load_effect(*args, **kwargs):
         self.mock_load_part.assert_has_calls([call1, call2], any_order=True)
 
 
+class ValidVersionTestCase(YamlTestCase):
+    scenarios = [
+        (version, dict(version=version)) for
+        version in ['buttered-popcorn', '1.2.3', 'v12.4:1:2~', 'HeLlo']
+    ]
+
+    def test_valid_yaml_invalid_version(self):
```
Comment (kalikiana) [inline snapcraft/tests/unit/project_loader/test_config.py]: This should probably be called `test_valid_yaml_valid_version`? Or even `test_valid_version`? I find these double valid/invalid rather confusing personally so I like the shorter name more.

File: snapcraft/tests/unit/project_loader/test_config.py [hunk 2]
```diff
@@ -909,6 +858,67 @@ def load_effect(*args, **kwargs):
...
+              part1:
+                plugin: nil
+        """).format(self.version))
+        _config.Config()
+
+
+class InvalidVersionTestCase(YamlTestCase):
+    scenarios = [
+        (version, dict(version=version)) for
+        version in ['*', '', '-bad', 'still-bad-', ':foo']
```
Comment (kalikiana) [inline snapcraft/tests/unit/project_loader/test_config.py]: Mind adding instances of the remaining special cases as well? That would be `.foo`, `+foo`, `~foo`, `_foo`, `bar.`, `bar:`, `bar-`, `bar_`.

File: snapcraft/tests/unit/project_loader/test_config.py (hunk 1 above)
Comment (kyrofa) [reply snapcraft/tests/unit/project_loader/test_config.py]: Yes these are silly indeed, thank you. Fixed.

File: snapcraft/tests/unit/project_loader/test_config.py (hunk 2 above)
Comment (kyrofa) [reply snapcraft/tests/unit/project_loader/test_config.py]: Sure, done.

**Pull request merged**